src/modules/voxelDensity/process.cpp

```diff
@@ -0,0 +1,104 @@
+#include "analyser/dataExporter.h"
+#include "analyser/dataOperator1D.h"
+#include "main/dissolve.h"
+#include "math/gaussFit.h"
+#include "module/context.h"
+#include "voxelDensity.h"
+#include <cmath>
+
+void VoxelDensityModule::addValue(Vec3<double> coords, double value, Array3D<double> &array)
+{
+    auto t = std::make_tuple((int)std::floor(coords.x * nAxisVoxels_), (int)std::floor(coords.y * nAxisVoxels_),
+                             (int)std::floor(coords.z * nAxisVoxels_));
+    array[t] += value;
+}
+
+Vec3<double> VoxelDensityModule::foldedCoordinates(Atom &atom, const Box *unitCell)
+{
+    atom.set(unitCell->foldFrac(atom.r()));
+    auto x = atom.x(), y = atom.y(), z = atom.z();
+    return Vec3<double>(x, y, z);
+}
+
+Module::ExecutionResult VoxelDensityModule::process(ModuleContext &context)
+{
+    auto &processingData = context.dissolve().processingModuleData();
+
+    // Calculate target property 3d map over unit cell voxels
+    auto array3D = processingData.realise<Array3D<double>>("Array3D", name());
+    array3D.initialise(nAxisVoxels_, nAxisVoxels_, nAxisVoxels_);
+
+    if (!restrictToSpecies_.empty())
+    {
+        auto cfgSpecies = targetConfiguration_->speciesPopulations();
+
+        for (const auto &[sp, N] : cfgSpecies)
+        {
+            bool included = false;
+
+            for (const auto &spTarget : restrictToSpecies_)
+            {
+                if (sp->name() == spTarget->name())
+                {
+                    included = true;
+                    break;
+                }
+            }
+
+            if (included)
+            {
+                targetConfiguration_->removeMolecules(sp);
+            }
+        }
+    }
+
+    auto unitCell = targetConfiguration_->box();
+    auto voxelVolume = unitCell->volume() / std::pow(nAxisVoxels_, 3);
+    auto atoms = targetConfiguration_->atoms();
```

Comment:
Here you're taking a copy of the entire atoms array which is going to be expensive and might cause issues with other object relationships.
```suggestion
    const auto &atoms = targetConfiguration_->atoms();
```